src/java.desktop/macosx/native/libosxapp/JNIUtilities.h

```diff
@@ -147,9 +154,44 @@
 
 /*********       EXCEPTION_HANDLING    *********/
 
+/*
+ * Some explanation to set context of the bigger picture.
+ * Before returning to Java from JNI, NSExceptions are caught - so long as
+ * the body of the native method is wrapped in the ENTER/EXIT macros.
+ * So if we want to directly return to Java from some nested Objective-C
+ * function when detecting a Java exception, we just need to raise an
+ * NSException. Then clear that right before returning to Java,
+ * leaving the Java exception to be seen back in Java-land.
+ *
+ * But if the current thread is the Appkit thread we might as well clear
+ * the Java Exception right now since there's nothing to receive it.
+ * In such a case control will propagate back to the run loop which might
+ * terminate the application. One drawback of that is that the location of
+ * termination does not show where the NSException originated.
+ * And for whatever reason, something swallows that exception.
+ * So as a debugging aid, when on the AppKit thread we can provide a
+ * way (via an env. var.) to log the location.
+ * Additionally provide a similar way to prevent the NSException being
+ * raised and instead just clear the Java Exception.
+ * Together these provide alternate behaviours for more debugging info
+ * or maybe a way for the app to continue running depending on the exact
+ * nature of the problem that has been detected and how survivable it is.
+ */
 #define CHECK_EXCEPTION() \
     if ((*env)->ExceptionOccurred(env) != NULL) { \
-        (*env)->ExceptionClear(env); \
+        if ([NSThread isMainThread] == YES) { \
```

Comment:
Isn't it is a similar code to the
https://github.com/openjdk/jdk/blob/67c221148159d94142a3a6d9ddadce2dff8c858b/src/java.desktop/macosx/native/libosxapp/NSApplicationAWT.m#L334
Where we just catch an exception log it, and continue to run our infinite loop? Why do we need to do something specific here? I mean we cannot drop any try/catch blocks anyway since an nsexception may be generated by some external system code.

Reply:
In my testing that code does not get called.  The new code does. So the old code could probably be deleted as useless but it is also harmless and just maybe it'll catch something else, even if it isn't doing anything that I can see.

Reply:
That code for sure should be called, it is even improved recently by JDK-8255681
I'll check how it was intended to work.

Reply:
I hadn't noticed that line - and definintely not realised it was recent.
I suppose he must have had a way to trigger it.
But I don't think it hurts to have both.

Reply:
I just tried to understand why we need to complicate this, to me, it is unclear why handling the same errors in the NSApplicationAWT.m does not work.

Reply:
Because of the reason I've said before. That logging in NSApplication AWT is not being seen. Something is swallowing it. So I'd say if anything remove that logging and keep the new one but as I also said it isn't harmful to have provision in case it also logs some case we aren't catching.